### PR TITLE
Move Embedding/ReversibleEmbedding to the quantization geometry protocol

### PR DESCRIPTION
**Pull request by JyotinderSingh**
## Description

`Embedding` returns a `LookupGeometry`. `ReversibleEmbedding` returns a `ReversibleLookupGeometry`. The two layers no longer define per-mode methods or override `quantized_build`. The int8 and int4 strategies implement the lookup family. The `reversible` flag of the geometry selects the reverse projection. The reverse projection reads the untied reverse table and its scale, or the transpose of the tied table. It then applies the logit soft cap. The soft cap and the embedding LoRA epilogue move to the shared helpers in `modes/common.py`.

The two layers move in one PR. The legacy per-mode methods of `ReversibleEmbedding` call `super()._int8_build`, `super()._int4_build`, `super()._int8_call` and `super()._int4_call`. A PR that moves only `Embedding` breaks the subclass.

### Reverse projection

The reverse projection now follows the forward table in every mode. This PR includes fixes to previously broken behavior in the following cases:

- The strategy quantizes an untied reverse table by the forward rule on its transpose. A user-supplied `weight_quantizer` applies to both tables (which previously failed for the reversed case). The default quantizer produces the same codes and scales as before.
- The quantized reverse projection of a tied layer adds the LoRA delta, as the float layer does (previously quantized variants ignored the delta).
- A quantized layer keeps the mask of its float forward pass when `mask_zero=True` (previously masks were wrongly ignored).
- The quantized reverse projection computes and returns its logits in `reverse_dtype` (the dtype was previously ignored).
- An int4 layer built from a dtype policy is weight-only on the reverse projection. int8 keeps its W8A8 default.
- The layer stores the reverse zero point as int8, like the forward zero point. Only grouped int4 checkpoints see this dtype change.
- The int4 forward lookup gathers the packed rows before it unpacks them. The cost scales with the batch size, not with the table size.

### Weight storage check

`Layer._strategy_owns_weight_storage()` reports whether the strategy created the weight storage in `quantized_build`. `Dense`, `EinsumDense`, `Embedding` and `ReversibleEmbedding` call it in `build`. The layers no longer look the strategy up themselves.

### Compatibility

Checkpoints and forward outputs of int8, int4 per-channel and float8 layers do not change.

### Tests

New tests cover:

- a custom `weight_quantizer` on both tables, tied and untied,
- the LoRA delta on the tied reverse projection,
- the mask after quantization,
- `reverse_dtype` on the quantized reverse projection,
- reverse outputs across a save and load round trip,
- the weight-only default of a policy-built int4 layer.

## Contributor Agreement

- [x] This PR is linked to an issue that has been assigned to me (see `Fixes #xxx` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/quantizers/modes/int4.py [hunk 1]
```diff
@@ -274,3 +281,241 @@ def _quantize_projection(self, layer, geometry, config):
...
+            )
+            # Transpose back to (input_dim, output_dim) layout
+            embeddings_value = ops.transpose(embeddings_value_t)
+            embeddings_scale = ops.transpose(scale_t)
+            embeddings_zero = ops.transpose(zero_t)
+
+        packed_embeddings_value, _, _ = pack_int4(embeddings_value, axis=-1)
+        return packed_embeddings_value, embeddings_scale, embeddings_zero
+
+    def _quantize_lookup(self, layer, geometry, config):
```
Comment (hertschuh) [inline keras/src/quantizers/modes/int4.py:498]: I'm confused, `_quantize_lookup` is never called.

File: keras/src/quantizers/modes/int8.py [hunk 2]
```diff
@@ -102,3 +109,106 @@ def _quantize_projection(self, layer, geometry, config):
...
+        weight_quantizer = QuantizationConfig.weight_quantizer_or_default(
+            config,
+            AbsMaxQuantizer(axis=-1),
+        )
+        embeddings_value, embeddings_scale = weight_quantizer(
+            weight, to_numpy=True
+        )
+        return embeddings_value, ops.squeeze(embeddings_scale, axis=-1), None
+
+    def _quantize_lookup(self, layer, geometry, config):
```
Comment (hertschuh) [inline keras/src/quantizers/modes/int8.py:197]: I'm confused, `_quantize_lookup` is never called.

File: keras/src/quantizers/modes/int4.py [hunk 3]
```diff
@@ -274,3 +281,241 @@ def _quantize_projection(self, layer, geometry, config):
         layer.kernel_scale.assign(kernel_scale)
         if kernel_zero is not None:
             layer.kernel_zero.assign(kernel_zero)
+
+    # --- Embeddings lookup (Embedding, ReversibleEmbedding) ---------------
+
+    def _build_lookup(self, layer, geometry, embeddings_shape, config):
```
Comment (hertschuh) [inline keras/src/quantizers/modes/int4.py:287]: I'm confused, `build_lookup` is never called.

File: keras/src/quantizers/modes/int8.py [hunk 4]
```diff
@@ -102,3 +109,106 @@ def _quantize_projection(self, layer, geometry, config):
         layer.quantized_build(kernel_shape, "int8", config)
         layer._kernel.assign(kernel_value)
         layer.kernel_scale.assign(kernel_scale)
+
+    # --- Embeddings lookup (Embedding, ReversibleEmbedding) ---------------
+
+    def _build_lookup(self, layer, geometry, embeddings_shape, config):
```
Comment (hertschuh) [inline keras/src/quantizers/modes/int8.py:115]: I'm confused, `build_lookup` is never called.

**Review by hertschuh — COMMENTED** [review]
I'm confused about any of this is wired.

How is `call` wired for instance?

**Comment by JyotinderSingh** [comment]
> I'm confused about any of this is wired.
> 
> How is `call` wired for instance?

We don't need to explicitly call any of these methods, they are wired automatically in the underlying geometry strategy base class: https://github.com/keras-team/keras/blob/5167f5dcaff85a1fdf44f6ef364e71d72e4afe9d/keras/src/quantizers/modes/common.py#L12-L46

File: keras/src/quantizers/geometry.py [hunk 5]
```diff
@@ -89,8 +98,14 @@ class QuantizationGeometry:
...
 
-    # Dispatch key for building, quantizing and the forward pass.
+    # Dispatch key for building and quantizing, and for the forward pass
+    # unless `call_family` overrides it.
     family = None
+    # Forward-pass dispatch key, when the forward pass needs a different
+    # implementation from build/quantize (a reversible lookup does).
+    call_family = None
+    # Whether the layer also projects back through its weight.
+    reversible = False
```
Comment (hertschuh) [inline keras/src/quantizers/geometry.py]: Do we need both `call_family` and `reversible`?

`call` uses `call_family`, but `build`, etc. uses the `reversible` flag.

Do we anticipate more `call_families` in the future?

Comment (JyotinderSingh) [reply keras/src/quantizers/geometry.py]: No, we don't need both. It was meant to serve as a future extension for new layers we may have added quantization support to (such as the several different kinds of conv layers) - but I guess it's too early to add such hooks. I've removed `call_family` and route via the `reversible` flag alone.

**Review by hertschuh — APPROVED** [review]
Thanks!
